packages/api/src/util/animal.js

```diff
@@ -142,3 +147,82 @@ export const upsertGroup = async (req, animalOrBatch, farm_id, trx) => {
     animalOrBatch.group_ids = [{ animal_group_id: group.id }];
   }
 };
+
+const getInvalidAnimalOrBatchIds = async (idsSet, model, farmId) => {
+  if (!idsSet.size) {
+    return [];
+  }
+
+  const invalidIdsSet = new Set(idsSet); // create a copy
+  const records = await model
+    .query()
+    .select('id')
+    .where({ farm_id: farmId, removal_date: null })
+    .whereIn('id', [...idsSet])
+    .whereNotDeleted();
+
+  records?.forEach(({ id }) => invalidIdsSet.delete(id));
+
+  return [...invalidIdsSet];
+};
+
+export const checkAnimalAndBatchIds = async (animalIds, batchIds, farmId, isRequired) => {
+  if (animalIds) {
+    checkIsArray(animalIds, 'animalIds');
+  }
+  if (batchIds) {
+    checkIsArray(batchIds, 'batchIds');
+  }
+
+  const animalIdsSet = new Set(animalIds);
+  const batchIdsSet = new Set(batchIds);
+
+  if (isRequired && !animalIdsSet.size && !batchIdsSet.size) {
+    throw customError('At least one of the animal IDs or animal batch IDs is required');
+  }
+
+  const invalidAnimalIds = await getInvalidAnimalOrBatchIds(animalIdsSet, AnimalModel, farmId);
+  const invalidBatchIds = await getInvalidAnimalOrBatchIds(batchIdsSet, AnimalBatchModel, farmId);
+
+  if (invalidAnimalIds.length || invalidBatchIds.length) {
+    throw customError(
+      'Some animal IDs or animal batch IDs do not exist, are removed, or are not associated with the given farm.',
+      400,
+      { invalidAnimalIds, invalidBatchIds },
+    );
+  }
+};
+
+export async function isOnOrAfterBirthAndBroughtInDates(date, animalIds, batchIds) {
```

Comment:
This function could be simplified by querying for the earliest date both for animals and batches, and then doing a comparison with those two dates only